platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/prometheus/AbstractMetricAdapter.java

```diff
@@ -16,19 +16,53 @@
 
 package com.swirlds.common.metrics.platform.prometheus;
 
+import static com.swirlds.common.metrics.platform.prometheus.NameConverter.fix;
+import static com.swirlds.logging.legacy.LogMarker.EXCEPTION;
+
+import com.swirlds.metrics.api.Metric;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import io.prometheus.client.SimpleCollector;
 import java.util.Objects;
 import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
 public abstract class AbstractMetricAdapter implements MetricAdapter {
-
+    private static final Logger log = LogManager.getLogger(AbstractMetricAdapter.class);
     protected final PrometheusEndpoint.AdapterType adapterType;
+    private final @NonNull CommonMetricValues values;
+
     private final AtomicInteger referenceCount = new AtomicInteger();
 
     /**
+     * @param metric the metric being adapted
      * @throws NullPointerException in case {@code adapterType} parameter is {@code null}
+     * @throws NullPointerException in case {@code metric} parameter is {@code null}
      */
-    protected AbstractMetricAdapter(final PrometheusEndpoint.AdapterType adapterType) {
+    protected AbstractMetricAdapter(
+            @NonNull final PrometheusEndpoint.AdapterType adapterType, @NonNull final Metric metric) {
+        this(adapterType, metric, false);
+    }
+
+    /**
+     * @param metric   the metric being adapted
+     * @param unitless if the {@code metric} being adapted represents a dimensionless metric. Prometheus will fail in
+     *                 case of setting a unit to an invalid metric type.
+     * @throws NullPointerException in case {@code adapterType} parameter is {@code null}
+     * @throws NullPointerException in case {@code metric} parameter is {@code null}
+     */
+    protected AbstractMetricAdapter(
+            @NonNull final PrometheusEndpoint.AdapterType adapterType,
+            @NonNull final Metric metric,
+            final boolean unitless) {
         this.adapterType = Objects.requireNonNull(adapterType, "adapterType must not be null");
+        this.values = new CommonMetricValues(unitless, Objects.requireNonNull(metric, "metric must not be null"));
+        this.values.checkAgainst(metric);
```

Comment:
At first glance this looked like a no-op because the call on line 59 would create the `values` object using the given `metric` object already. However, I see that the implementation of the `CommonMetricValues` constructor actually calls `fix()` on each of the individual values, while the `checkAgainst` uses a simple Objects.equals() to compare the individual values.

So essentially, in these two lines of code we're trying to assert that `fix()` returns its argument verbatim, or else we log a bunch of errors.

This brings the question as to
1. whether this is the intended purpose of this code,
2. and/or whether the calls to `fix()` are required (because this check will fail the moment the `fix()` returns anything different from its input argument),
3. and if they are required, then shouldn't the checkAgainst() take the `fix()` into account?

Could you please clarify?

Reply:
Sure, the "fix()" method is the existing mechanism to replace a metric name component (category/name/unit), so it is valid against some predefined naming rules Prometheus requires. If the result of invoking fix() for a string returns the exact string, the value is a valid name for Prometheus.   
The requirement is that we want to avoid/identify when the metric name is changed under the hood by the framework and inform developers that they should adjust the metric name. That is why it compares the actual parameter vs. the value after being #fix()ed.
Let me rethink the design so it is more explicit; I see how having a check method that only performs log statements collaborates with the idea that something must be wrong.
The initial design had all the code in the constructor and exposed the values after being fixed to the subclasses, but that seemed weak encapsulation.

Reply:
Got it. Thanks for the clarification. This makes sense. I'd suggest to add a code comment (either in checkAgainst() javadoc, or above the call to the method) to briefly describe the intention of the check. This would make the code a lot clearer.

Reply:
@anthony-swirldslabs Let me know if the logic + name change + additional javadoc did a better job expressing the intention

Reply:
That looks perfect! Thank you!